File: 1329-sort-the-matrix-diagonally/1329-sort-the-matrix-diagonally.cpp
class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        int n=mat.size();
        int m=mat[0].size();
        int i=n-1,j=0;
        int cnt=0;
        while(i>=0){
            j=0;
            vector<int> v;
            for(int k=i;k<n and j<m;k++,j++){
                v.push_back(mat[k][j]);
            }
            
            sort(v.begin(),v.end());
            int l=0;
            j=0;
            for(int k=i;k<n and j<m;k++,j++){
                mat[k][j]=v[l];
                l++;
            }
            i--;
        }
        
        while(++cnt<m){
            j=cnt;
            i=0;
            vector<int> v;
            for(;i<n and j<m; i++,j++){
                v.push_back(mat[i][j]);
            }
            
            sort(v.begin(),v.end());
            i=0,j=cnt;
            int l=0;
            
            for(;i<n and j<m;i++,j++){
                mat[i][j]=v[l];
                l++;
            }
            
        }
        
        
        return mat;
    }
};

File: 1329-sort-the-matrix-diagonally/1329-sort-the-matrix-diagonally.java
class Solution {
	public int[][] diagonalSort(int[][] mat) {
		//sort row-wise diagonally [00, 10, 20]
		for(int i=0; i<mat.length; i++) countAndSort(mat, i, 0);

		//sort column-wise diagonally [01, 02, 03]
		for(int i=0; i<mat[0].length; i++) countAndSort(mat, 0, i);

		return mat;
	}

	public void countAndSort(int[][] mat, int i, int j) {
		int[] arr = new int[101];
		int r = i;
		int c = j;

		//store the count in arr
		while(r < mat.length && c < mat[0].length)
			arr[mat[r++][c++]]++;

		r = i;
		c = j;

		//loop through [1-100] constraint range and update mat diagonally
		for(int k=1; k<101; k++) {
			while(arr[k]-- > 0) {
				mat[r++][c++] = k;
			}
		}
	}
}